### PR TITLE
[SYCL] Add/use is_device_copyable<> type trait class

### PR DESCRIPTION
**Pull request by v-klochkov**
The new type trait class is_device_copyable added to SYCL header files
replaces the checks done in C++FE previously.

The type trait can get specializations for user data types to
enable copying of those classes from HOST to DEVICE or between DEVICEs.
Such type must follow the restrictions listed in SYCL2020 SPEC for devices
copyable types.

Currently the default behavior of is_device_copyable is equal to:
  (is_trivially_copyable_v ||
  (is_trivially_copy_constructible_v && is_trivially_destructible_v))

That is done for compatibility with the current SYCL behavior that allowed
passing trivially copy-constructible/destructible types between devices.
The warning for non-trivially copyable but constructible/destructible types
is implemented but is not printed, which will be fixed separately.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

File: sycl/test/basic_tests/is_device_copyable.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// RUN: not %clangxx -fsycl -fsycl-device-only -fsyntax-only -Xclang -verify \
+// RUN: %s -I %sycl_include 2>&1 | FileCheck -check-prefix=CHECK-ERR %s
```
Comment (romanovvlad) [inline sycl/test/basic_tests/is_device_copyable.cpp:2]: Not sure I understand why the test is designed this way, could you please elaborate?
Can we just use `expected-warning` and `expected-error` comments for clang verify instead of using `FileCheck`?
Also, instead of marking some types as device copyable in a separate compilation can we just add duplicate types and have one compilation?
can we introduce additional types which will be

**Review by AaronBallman — APPROVED** [review]
The FE bits LGTM (and the SYCL library bits also look good).

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: expected-error is what I tried first, but that did not work as the error line number points to header, but 'expected-error' needs to list the line number where it is placed in is_device_copyable.cpp file, e.g.: expected-error@+1

Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: > Also, instead of marking some types as device copyable in a separate compilation can we just add duplicate types and have one compilation?

Yes, I'll do that.

Comment (AaronBallman) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: > expected-error is what I tried first, but that did not work as the error line number points to header, but 'expected-error' needs to list the line number where it is placed in is_device_copyable.cpp file, e.g.: expected-error@+1

FWIW, there are a bunch of features to the diagnostic verifier that can help with this: https://clang.llvm.org/doxygen/classclang_1_1VerifyDiagnosticConsumer.html (see the Detailed Description part).

So you can likely use something like `// expected-error@foo.h:15 {{whatever}}` to check the diagnostic instead.

Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: Yes, that is doable, but the line in that header may shift when unrelated changes are done there.

Comment (AaronBallman) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: There are more features to help with that. :-) You can create a "bookmark" that is used to match diagnostics. e.g.,
```
// Header.h
int whatever; // #SuperAwesomeError

// Source.cpp
foo(); // expected-error@#SuperAwesomeError
```
This pairs the expected error with the bookmark defined in the header with the special comment.

Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: Right. Adding a bookmark in sycl/include/CL/sycl/types.hpp does not seem reasonable.

Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: Fixed the test by adding duplicated types and removing one of compilation lines: 302ab3e79524cc9dc935af18e3949960470729f5

File: sycl/include/CL/sycl/types.hpp [hunk 2]
```diff
@@ -2274,6 +2274,100 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
...
+};
+
+template <typename T> struct CheckBasesAreDeviceCopyable<T, 0> {};
+
+// All the captures of a lambda or functor of type FuncT passed to a kernel
+// must be is_device_copyable, which extends to bases and fields of FuncT.
+// Fields are captures of lambda/functors and bases are possible base classes
+// of functors also allowed by SYCL.
+// The SYCL-2020 implementation must check each of the fields & bases of the
+// type FuncT, only one level deep, to see if they are all device copyable
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/CL/sycl/types.hpp]: Because one level deep is enough. I updated the comment: 302ab3e79524cc9dc935af18e3949960470729f5

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 1 above)
Comment (AaronBallman) [reply sycl/test/basic_tests/is_device_copyable.cpp:2]: > Right, but adding a bookmark in sycl/include/CL/sycl/types.hpp does not seem reasonable.

Ah, I think I've been caught out yet again by SYCL abusing lit tests a little bit. In lit, all the include files are part of the test set and so the expectation is that you can add your diagnostic checks to them. SYCL elected to do full end-to-end testing instead of something more cross-system-stable. :-(

For your patch, the current approach seems fine to me.

**Review by romanovvlad — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -1606,26 +1606,6 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
...
-               << FieldTy;
-
-    if (!FieldTy->isStructureOrClassType())
-      return false;
-
-    CXXRecordDecl *RD =
-        cast<CXXRecordDecl>(FieldTy->getAs<RecordType>()->getDecl());
-    if (!RD->hasTrivialCopyConstructor())
-      return Diag.Report(FD->getLocation(),
-                         diag::err_sycl_non_trivially_copy_ctor_dtor_type)
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1622]: Please remove both diagnostics from DiagnosticSemaKinds.td as they no longer appear to be used.

Comment (v-klochkov) [reply clang/lib/Sema/SemaSYCL.cpp:1622]: Sure, done here: 7852670026636363f427958f3712800d6f7bc2e0

File: sycl/test/basic_tests/is_device_copyable.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    int D = IamAlsoBadButCopyable.i;
+
+    int E = IamAlsoGood.i;
+    int F = IamAlsoBad.i;
+  });
+}
+// CHECK-ERR: no expected directives found
+// CHECK-ERR-NEXT: diagnostics seen but not expected
+// CHECK-ERR-NEXT: is_device_copyable<B, void>
+// CHECK-ERR-NEXT: is_device_copyable<D, void>
```
Comment (AaronBallman) [inline sycl/test/basic_tests/is_device_copyable.cpp:70]: The diagnostics generated are telling you something important -- you're not testing what you think you're testing and this output is bad.

`no expected directives found` is coming from passing `-verify` on the RUN line but not telling the diagnostic verifier what diagnostics you expected to find. `diagnostics seen but not expected` is the diagnostic verifier telling you that there were actually diagnostics found.

I *think* you can remove the use of `FileCheck` and instead put the correct `// expected-error {{}}` markings in the file and let the diagnostic verifier handle checking the test results (preferred approach). Alternatively, you can remove the `-verify` and add the correct `CHECK` lines (only do this if you can't get `-verify` to work).

File: sycl/test/basic_tests/is_device_copyable.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  int i;
+  ~D();
+};
+
+// Identical to D, but will be declared as device copyable.
+struct DCopyable {
+  int i;
+  ~DCopyable();
+};
+
```
Comment (AaronBallman) [inline sycl/test/basic_tests/is_device_copyable.cpp:43]: This is a good start to testing, but only tests some of the super common stuff. Please add tests for *all* of the ways that checking device copyability can fail (like base classes), as well as edge case scenarios (classes with no members or bases, virtual bases with difficult inheritance patterns, classes with noncopyable static members (as those don't get copied as part of the object), bit-fields, and whatever other odd situations you can come up with.

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:70]: @AaronBallman, haven't we agreed that "// expected-error {{}}" is not usable for this test, and the current approach is fine: https://github.com/intel/llvm/pull/3905#discussion_r649383664
This approach is also not invented by me, I've seen it used somewhere in clang or llvm.

> Alternatively, you can remove the -verify and add the correct CHECK lines (only do this if you can't get -verify to work

I don't understand the alternative idea. IMO, the current approach is reasonable and does not miss any cases where a) diagnistic is expected but not printed, b) when diagnostic is not expected but printed.

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:43]: This test is rework of the deleted clang test that is deleted in this PR: https://github.com/intel/llvm/pull/3905/files#diff-75331d55b17d64c7f82af56c111baced01571af4ef7eef17678e3bb730e2dc08
I'll add more tests for sure.

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 4 above)
Comment (AaronBallman) [reply sycl/test/basic_tests/is_device_copyable.cpp:70]: Good point, we have (thank you for the reminder), but that doesn't mean the test isn't incorrect. Some of those `CHECK` lines are reporting errors due to the incorrect use of `-verify`, and that's what I'd like to see fixed. e.g., https://godbolt.org/z/6Eb7d66j5

The alternative idea is: remove `-verify` from the RUN line and remove the two CHECK lines that will stop showing up.

File: sycl/include/CL/sycl/types.hpp [hunk 6]
```diff
@@ -2279,6 +2279,100 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
...
+template <typename T, typename... Ts>
+struct is_device_copyable<std::tuple<T, Ts...>>
+    : detail::bool_constant<is_device_copyable<T>::value &&
+                            is_device_copyable<std::tuple<Ts...>>::value> {};
+
+namespace detail {
+template <typename T, typename = void>
+struct IsDeprecatedDeviceCopyable : std::false_type {};
+
+// TODO: using C++ attrubute [[deprecated]] or the macro __SYCL2020_DEPRECATED
```
Comment (AaronBallman) [inline sycl/include/CL/sycl/types.hpp:2322]:
```suggestion
// TODO: using C++ attribute [[deprecated]] or the macro __SYCL2020_DEPRECATED
```

File: sycl/include/CL/sycl/types.hpp [hunk 7]
```diff
@@ -2279,6 +2279,100 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
...
+#define SYCL_DEVICE_COPYABLE 1
+
+/// is_device_copyable is a user specializable class template to indicate
+/// that a type T is device copyable, which means that SYCL implementation
+/// may copy objects of the type T between host and device or between two
+/// devices.
+/// Specializing is_device_copyable such a way that
+/// is_device_copyable_v<T> == true on a T that does not satisfy all
+/// the requirements of a device copyable type may result into undefined
+/// behavior.
```
Comment (AaronBallman) [inline sycl/include/CL/sycl/types.hpp:2294]:
```suggestion
/// the requirements of a device copyable type is undefined behavior.
```

File: sycl/include/CL/sycl/types.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/types.hpp]: Fixed in https://github.com/intel/llvm/pull/3905/commits/ace3b783d59af3b699f9d8a0e0989a413e836b8c

File: sycl/include/CL/sycl/types.hpp (hunk 7 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/types.hpp]: Fixed in https://github.com/intel/llvm/pull/3905/commits/ace3b783d59af3b699f9d8a0e0989a413e836b8c

File: sycl/test/basic_tests/is_device_copyable.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+// Not copyable type, but it will be declared as device copyable.
+struct BCopyable {
+  int i;
+  BCopyable(int _i) : i(_i) {}
+  BCopyable(const BCopyable &x) : i(x.i) {}
+};
+
+struct C : A {
+  const A C2;
```
Comment (AaronBallman) [inline sycl/test/basic_tests/is_device_copyable.cpp:28]: It'd be useful to have some comments around here explaining that this type is not trivially copyable but it is trivially copy constructible and so this hits the deprecated functionality (right?).

File: sycl/test/basic_tests/is_device_copyable_neg.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+  FunctorB FB;
+  Q.single_task<class TestC>(FB);
+}
+
+// CHECK-ERR: static_assert failed due to requirement 'is_device_copyable<A,
+// void> CHECK-ERR: static_assert failed due to requirement
+// 'is_device_copyable<B, void> CHECK-ERR: static_assert failed due to
+// requirement 'is_device_copyable<C, void> CHECK-ERR: static_assert failed due
+// to requirement 'is_device_copyable<D, void>
```
Comment (AaronBallman) [inline sycl/test/basic_tests/is_device_copyable_neg.cpp:75]:
```suggestion
// CHECK-ERR: static_assert failed due to requirement 'is_device_copyable<A, void>
// CHECK-ERR: static_assert failed due to requirement 'is_device_copyable<B, void>
// CHECK-ERR: static_assert failed due to requirement 'is_device_copyable<C, void>
// CHECK-ERR: static_assert failed due to requirement 'is_device_copyable<D, void>
```
You should also add the line number information to the CHECK-ERR lines so it's clear where in the test file the diagnostics are issued.

**Review by AaronBallman — COMMENTED** [review]
Just checking: are you planning to add additional test coverage for more esoteric situations? Things like virtual inheritance, inheriting from a lambda (as done for `std::variant` visitation), that sort of thing?

File: sycl/test/basic_tests/is_device_copyable.cpp (hunk 8 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable.cpp:32]: Added in https://github.com/intel/llvm/pull/3905/commits/da49819bce2df28310ba8af73934aa87fcf88ce6

File: sycl/test/basic_tests/is_device_copyable_neg.cpp (hunk 9 above)
Comment (v-klochkov) [reply sycl/test/basic_tests/is_device_copyable_neg.cpp]: That mess was created by clang-format. Fixed it and also added source lines numbers in https://github.com/intel/llvm/pull/3905/commits/da49819bce2df28310ba8af73934aa87fcf88ce6
Each of types B, C, D is used only once; C and D are copies of B to make the diagnostic/error checks look more obvious.

**Comment by v-klochkov** [comment]
> Just checking: are you planning to add additional test coverage for more esoteric situations? Things like virtual inheritance, inheriting from a lambda (as done for `std::variant` visitation), that sort of thing?

Yes, this patch is the first of 3 planned patches. It included only basic compile time checks.
The 2nd patch will add support for implicitly device copyable types and 3rd will add more tests to intel/llvm-test-suite that would be verifying more situations in runtime (on CPU/GPU/etc).

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by AaronBallman** [comment]
> > Just checking: are you planning to add additional test coverage for more esoteric situations? Things like virtual inheritance, inheriting from a lambda (as done for `std::variant` visitation), that sort of thing?
> 
> Yes, this patch is the first of 3 planned patches. It included only basic compile time checks.
> The 2nd patch will add support for implicitly device copyable types and 3rd will add more tests to intel/llvm-test-suite that would be verifying more situations in runtime (on CPU/GPU/etc).

Thanks for letting me know!

File: sycl/include/CL/sycl/types.hpp [hunk 10]
```diff
@@ -2279,6 +2279,99 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
...
+struct IsDeprecatedDeviceCopyable : std::false_type {};
+
+// TODO: using C++ attribute [[deprecated]] or the macro __SYCL2020_DEPRECATED
+// does not produce expected warning message for the type 'T'.
+template <typename T>
+struct __SYCL2020_DEPRECATED("This type isn't device copyable in SYCL 2020")
+    IsDeprecatedDeviceCopyable<
+        T, std::enable_if_t<std::is_trivially_copy_constructible<T>::value &&
+                            std::is_trivially_destructible<T>::value &&
+                            !is_device_copyable<T>::value>> : std::true_type {};
```
Comment (cperkinsintel) [inline sycl/include/CL/sycl/types.hpp:2328]: Am I understanding this correctly? We are saying an object `IsDeprecatedDeviceCopyable` if it is `trivially_copy_constructible` and `trivially_destructible` but NOT `trivially_copyable`  ?  If that's what we want to say, maybe we should use `!std::is_trivially_copyable`  so it's clearer.   
But if you wanted this to be "in the case where we are not device_copyable",  then the old way would suffice. But in that case, the OR checks will capture that.  If that is the desire, then maybe leave this third clause out of the conjunction.  
Please correct me if I misunderstood.

Comment (v-klochkov) [reply sycl/include/CL/sycl/types.hpp:2328]: > We are saying an object IsDeprecatedDeviceCopyable if it is trivially_copy_constructible and trivially_destructible but NOT trivially_copyable

That is exactly what is needed and meant by this code. If type is NOT trivially copyable, but is constructible/destructible, then it is allowed to be passed from HOST to DEVICE, but as deprecated behavior (to stay compatible with older SYCL std).

Comment (cperkinsintel) [reply sycl/include/CL/sycl/types.hpp:2328]: Then I'd recommend using `!std::is_trivially_copyable` so it's easier to understand, and not tied to `is_device_copyable` in case that were to change in the future.

**Review by cperkinsintel — APPROVED** [review]

File: sycl/include/CL/sycl/types.hpp (hunk 10 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/types.hpp:2328]: is_device_copyable includes 2 situations: a) is_trivially_device_copyable, b) situation when user declared some type as device_copyable. Both of (a) and (b) are not considered deprecated pass for copying data from host to device. (a) and (b) are good ways (acc to SYCL 2020) to allow type copying.

**Comment by v-klochkov** [comment]
@turinevgeny , please review/approve the changes, your review is required before proceeding to merge.

File: sycl/include/CL/sycl/types.hpp (hunk 10 above)
Comment (cperkinsintel) [reply sycl/include/CL/sycl/types.hpp:2328]: But in this case, wouldn't the  `is_device_copyable<T> || isDeprecatedDeviceCopyable<T>` capture that? It works even if `isDeprecatedDeviceCopyable` does not contain `!is_device_copyable`.    And that is the only use for `isDeprecatedDeviceCopyable`, to appear in those OR clauses. 
 Anyway, doing all these template substitutions and boolean logic in my head hurts. If only I had a machine that could do it for me.... ;-)

**Review by turinevgeny — APPROVED** [review]
